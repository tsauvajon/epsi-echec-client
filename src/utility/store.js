import { createStore, applyMiddleware } from 'redux';
import createSocketIoMiddleware from 'redux-socket.io';
import io from 'socket.io-client';
import checksApp from '../reducers';
import { PieceEnum } from './util';
import initialGame from './initialGame';

// définit la couleur de la pièce
const definePlayer = (i) => {
  if (i < 16) {
    return 'black';
  } else if (63 - i < 16) {
    return ('white');
  }

  return undefined;
};

// on définit le type de la pièce (grâce à la symétrie)
// 63-x = symétrie centrale
const definePiece = (i) => {
  // les rois et dames ne sont pas symétriques
  switch (i) {
    case 3:
    case 59:
      return PieceEnum.QUEEN;
    case 4:
    case 60:
      return PieceEnum.KING;
    default: break;
  }

  const half = (j) => {
    if (j >= 8 && j <= 15) {
      return PieceEnum.PAWN;
    }

    switch (j) {
      case 0:
      case 7 - 0:
        return PieceEnum.ROOK;
      case 1:
      case 7 - 1:
        return PieceEnum.KNIGHT;
      case 2:
      case 7 - 2:
      default:
        return PieceEnum.BISHOP;
    }
  };

  if (i < 16) {
    return half(i);
  } else if (63 - i < 16) {
    return half(63 - i);
  }
  return undefined;
};

// crée le board par défaut (placement initial)
/*
  TODO : refactor avec array.filter() ??; ex:  [2, 4, 6].filter(gt4).map(double); [8, 12]
*/
const getDefaultPieces = () => (
  // on déclare un Array de 64 élements
  // puis on applique les fonctions de définition sur chacun des squares
  [...Array(64)].map(
    (_, i) => ({
      id: i,
      player: definePlayer(i),
      piece: definePiece(i),
    }),
  )
);

const initialState = {
  squares: getDefaultPieces(),
  game: initialGame,
};

// Initialisation de socket io

const socket = io('http://localhost:3000');
const socketIoMiddleware = createSocketIoMiddleware(socket, 'server/');
const store = createStore(checksApp, initialState, applyMiddleware(socketIoMiddleware));

store.subscribe(() => {
  // console.log('new client state', store.getState());
});

// store.dispatch({
//   type: 'ADD_PIECE',
//   id: 17,
//   player: 'black',
//   piece: 'PAWN',
// });

export default store;
