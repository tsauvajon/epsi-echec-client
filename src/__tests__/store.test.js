import store from '../utility/store';

it('generates default store', () => {
  expect(store.getState()).toMatchSnapshot();
});
