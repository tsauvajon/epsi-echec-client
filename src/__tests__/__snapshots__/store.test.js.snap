// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates default store 1`] = `
Object {
  "game": Object {
    "castling": Array [
      0,
      7,
      56,
      63,
    ],
    "nextPlayer": "white",
    "selected": null,
  },
  "squares": Array [
    Object {
      "id": 0,
      "piece": "rook",
      "player": "black",
    },
    Object {
      "id": 1,
      "piece": "knight",
      "player": "black",
    },
    Object {
      "id": 2,
      "piece": "bishop",
      "player": "black",
    },
    Object {
      "id": 3,
      "piece": "queen",
      "player": "black",
    },
    Object {
      "id": 4,
      "piece": "king",
      "player": "black",
    },
    Object {
      "id": 5,
      "piece": "bishop",
      "player": "black",
    },
    Object {
      "id": 6,
      "piece": "knight",
      "player": "black",
    },
    Object {
      "id": 7,
      "piece": "rook",
      "player": "black",
    },
    Object {
      "id": 8,
      "piece": "pawn",
      "player": "black",
    },
    Object {
      "id": 9,
      "piece": "pawn",
      "player": "black",
    },
    Object {
      "id": 10,
      "piece": "pawn",
      "player": "black",
    },
    Object {
      "id": 11,
      "piece": "pawn",
      "player": "black",
    },
    Object {
      "id": 12,
      "piece": "pawn",
      "player": "black",
    },
    Object {
      "id": 13,
      "piece": "pawn",
      "player": "black",
    },
    Object {
      "id": 14,
      "piece": "pawn",
      "player": "black",
    },
    Object {
      "id": 15,
      "piece": "pawn",
      "player": "black",
    },
    Object {
      "id": 16,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 17,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 18,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 19,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 20,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 21,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 22,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 23,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 24,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 25,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 26,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 27,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 28,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 29,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 30,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 31,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 32,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 33,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 34,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 35,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 36,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 37,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 38,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 39,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 40,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 41,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 42,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 43,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 44,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 45,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 46,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 47,
      "piece": undefined,
      "player": undefined,
    },
    Object {
      "id": 48,
      "piece": "pawn",
      "player": "white",
    },
    Object {
      "id": 49,
      "piece": "pawn",
      "player": "white",
    },
    Object {
      "id": 50,
      "piece": "pawn",
      "player": "white",
    },
    Object {
      "id": 51,
      "piece": "pawn",
      "player": "white",
    },
    Object {
      "id": 52,
      "piece": "pawn",
      "player": "white",
    },
    Object {
      "id": 53,
      "piece": "pawn",
      "player": "white",
    },
    Object {
      "id": 54,
      "piece": "pawn",
      "player": "white",
    },
    Object {
      "id": 55,
      "piece": "pawn",
      "player": "white",
    },
    Object {
      "id": 56,
      "piece": "rook",
      "player": "white",
    },
    Object {
      "id": 57,
      "piece": "knight",
      "player": "white",
    },
    Object {
      "id": 58,
      "piece": "bishop",
      "player": "white",
    },
    Object {
      "id": 59,
      "piece": "queen",
      "player": "white",
    },
    Object {
      "id": 60,
      "piece": "king",
      "player": "white",
    },
    Object {
      "id": 61,
      "piece": "bishop",
      "player": "white",
    },
    Object {
      "id": 62,
      "piece": "knight",
      "player": "white",
    },
    Object {
      "id": 63,
      "piece": "rook",
      "player": "white",
    },
  ],
}
`;
